core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterManager.java

```diff
@@ -0,0 +1,105 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.trino.operator.exchange;
+
+import io.airlift.log.Logger;
+import io.airlift.units.DataSize;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Supplier;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This class is separated from {@link ScaleWriterExchanger} to ensure that the same state
+ * is used across all local splits.
+ */
+@ThreadSafe
+public class ScaleWriterManager
+{
+    private static final Logger log = Logger.get(ScaleWriterExchanger.class);
+
+    private final LocalExchangeMemoryManager memoryManager;
+    private final int totalWriters;
+    private final long bufferSizeMinScalingThreshold;
+    private final Supplier<Long> physicalWrittenBytesSupplier;
+    private final long writerMinSize;
+    private final Runnable setWritersFullyScaled;
+    private final AtomicBoolean isSetWritersFullyScaledExecuted = new AtomicBoolean(false);
+    // start with single writer and increase the writer count based on
+    // physical written bytes and buffer utilization.
+    private final AtomicInteger writerCount = new AtomicInteger(1);
+    private final AtomicLong lastScaleUpPhysicalWrittenBytes = new AtomicLong(0);
+
+    public ScaleWriterManager(
+            LocalExchangeMemoryManager memoryManager,
+            int totalWriters,
+            long maxBufferedBytes,
+            Supplier<Long> physicalWrittenBytesSupplier,
+            DataSize writerMinSize,
+            Runnable setWritersFullyScaled)
+    {
+        this.memoryManager = requireNonNull(memoryManager, "memoryManager is null");
+        this.totalWriters = totalWriters;
+        // We are aiming to keep the buffer size below 50% of its limit
+        this.bufferSizeMinScalingThreshold = maxBufferedBytes / 2;
+        this.physicalWrittenBytesSupplier = requireNonNull(physicalWrittenBytesSupplier, "physicalWrittenBytesSupplier is null");
+        this.writerMinSize = requireNonNull(writerMinSize, "writerMinSize is null").toBytes();
+        this.setWritersFullyScaled = requireNonNull(setWritersFullyScaled, "setWritersFullyScaled is null");
+    }
+
+    public int getWriterCount()
+    {
+        updateWriterCount(memoryManager.getBufferedBytes());
+        return writerCount.get();
+    }
+
+    private void updateWriterCount(long bufferedBytes)
+    {
+        // Scale up writers when current buffer memory utilization is more than 50% of the
+        // maximum and physical written bytes by the last scaled up writer is greater than
+        // writerMinSize.
+        if (writerCount.get() < totalWriters && bufferedBytes >= bufferSizeMinScalingThreshold) {
```

Comment:
does the condition for `bufferedBytes` mean that we want to keep using the same writer if it can keep up with incoming data?
If so, what do you think about expressing this in the comment more clearly?